src/transformers/generation/continuous_batching/cache.py

```diff
@@ -13,26 +13,110 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 from collections import deque
-from math import floor, sqrt
+from math import floor, gcd, sqrt
 from typing import Optional, Union
 
 import torch
 
 from ...configuration_utils import PretrainedConfig
 from ...generation.configuration_utils import GenerationConfig
 from ...utils.metrics import attach_tracer, traced
-from .classes import RequestState, get_device_and_memory_breakdown, logger
+from .cache_manager import CacheManager, FullAttentionCacheManager, SlidingAttentionCacheManager
+from .requests import get_device_and_memory_breakdown, logger
+
+
+NO_SLIDING_WINDOW = 1
+
+
+def group_layers_by_attn_type(config: PretrainedConfig) -> tuple[list[list[int]], list[str]]:
+    """
+    Group layers depending on the attention mix, according to VLLM's hybrid allocator rules:
+        - Layers in each group need to have the same type of attention
+        - All groups have the same number of layers
+
+    For a model with the following layer types: ["sliding", "full", "full", "sliding"]
+    We would get two groups: [0, 3] and [1, 2].
```

Comment:
we need an example that makes use of the group size

Reply:
["sliding", "full", "full", "sliding", "full", "full", "full", "full"]
= [0, 3] and [1, 2], [4,5], [6,7]